### PR TITLE
mgr/dashboard: improve the kcli bootstrap process

### PR DESCRIPTION
**Pull request by nizamial09**
I have a new script added for starting the kcli cluster called quick-bootstrap.sh

The goal is to use that script to download the ceph image on local (rather than inside vm) and then copy them over to all the vms that is being spawned by the kcli. This way all the hosts will get the ceph image which will make the deployment loads faster.

Another thing I added is to add some dnf.conf to improve parallel_downlaods and get the fastest server to install deps

eg:
```
╰─$ ./quick-bootstrap.sh -h                                                                                    255 ↵
+ set +x
Usage: ./quick-bootstrap.sh [OPTIONS]

Options:
  -u, --use-cached-image     Uses the existing podman image in local. Only use this if there is such an image present.
  -dir, --ceph-dir             Use this to provide the local ceph directory. eg. --ceph-dir=/path/to/ceph
  -e, --expanded-cluster     To add all the hosts and deploy OSDs on top of it.
  -h, --help             Display this help message.

Example:
  ./quick-bootstrap.sh --use-cached-image

```

```
./quick-bootstrap.sh -u --ceph-dir=/home/nia/projects/ceph
```

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/pybind/mgr/dashboard/ci/cephadm/quick-bootstrap.sh [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    podman pull ${image_name}
+fi
+
+rm -f ceph_image.tar
+
+printf "Saving the image: %s\n" "$image_name"
+podman save -o ceph_image.tar quay.ceph.io/ceph-ci/ceph:main
+
+printf "Creating the plan\n"
+kcli create plan -f ceph_cluster.yml -P ceph_dev_folder=/home/nia/projects/ceph -P quick_install=True ceph
```
Comment (Pegonzal) [inline src/pybind/mgr/dashboard/ci/cephadm/quick-bootstrap.sh:54]:
```suggestion
kcli create plan -f ceph_cluster.yml -P ceph_dev_folder=${from_config}/projects/ceph -P quick_install=True ceph
```

File: src/pybind/mgr/dashboard/ci/cephadm/quick-bootstrap.sh [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+attempt=0
+
+MAX_ATTEMPTS=15
+SLEEP_INTERVAL=5
+
+printf "Waiting for the host to be reachable\n"
+while [[ $attempt -lt $MAX_ATTEMPTS ]]; do
+    if ssh -o StrictHostKeyChecking=no -o BatchMode=yes -o ConnectTimeout=10 root@192.168.100.100 exit; then
```
Comment (Pegonzal) [inline src/pybind/mgr/dashboard/ci/cephadm/quick-bootstrap.sh:63]: Should the IP be taken from config? 
```suggestion
    if ssh -o StrictHostKeyChecking=no -o BatchMode=yes -o ConnectTimeout=10 root@${fip_rom_config} exit; then
```

**Review by Pegonzal — COMMENTED** [review]
Thank you @nizamial09!

Looks promising, I'll try to take a better look and test it later on, left you a couple comments for now of what I saw:

File: src/pybind/mgr/dashboard/ci/cephadm/ceph_cluster.yml [hunk 3]
```diff
@@ -36,7 +36,18 @@ parameters:
  files:
   - bootstrap-cluster.sh
  cmds:
+ # updating the dnf.conf to make the dnf faster
+ - echo -e "[main]\nfastestmirror=true\nmax_parallel_downloads=10\nmetadata_expire=1h\nclean_requirements_on_remove=true\nassumeyes=true\ngpgcheck=1\nkeepcache=0\nplugins=1\ninstallonly_limit=3" | sudo tee /etc/dnf/dnf.conf
```
Comment (epuertat) [inline src/pybind/mgr/dashboard/ci/cephadm/ceph_cluster.yml:40]: Let's try to make things more readable when possible (yaml [supports multiple multiline syntaxes](https://yaml-multiline.info/)).
```suggestion
 - |
   echo "\
[main]
fastestmirror=true
max_parallel_downloads=10
metadata_expire=1h
...
```
Or even better, add this as separate file to the `files` section. Embedding different languages (Jinja + YAML + bash + systemd.conf) makes this hard to follow.

File: src/pybind/mgr/dashboard/ci/cephadm/bootstrap-cluster.sh [hunk 4]
```diff
@@ -12,6 +12,26 @@ CEPHADM_SRC="/mnt/{{ ceph_dev_folder }}/src/cephadm/cephadm"
...
+)
+
+declare -A config_sections=(
+  ["mon_allow_pool_size_one"]="mon"
+  ["mon_allow_pool_delete"]="mon"
+  ["osd_pool_default_min_size"]="global"
+  ["osd_pool_default_size"]="global"
+  ["mon_data_avail_crit"]="mon"
+  ["mon_data_avail_warn"]="mon"
+)
```
Comment (epuertat) [inline src/pybind/mgr/dashboard/ci/cephadm/bootstrap-cluster.sh:33]: `cephadm bootstrap` supports a `--config <config.file>` to import a basic config file. I'd rather add that as an additional file and copy to the cluster through the `- files` YAML key:
```ini
[global]
osd_pool_default_min_size=1
osd_pool_default_size=1
...
```

When things look overcomplicated, probably there's a simpler way to do it.

File: src/pybind/mgr/dashboard/ci/cephadm/ceph_cluster.yml [hunk 5]
```diff
@@ -36,7 +36,18 @@ parameters:
...
  - dnf -y install python3 chrony lvm2 podman
+ # setting up an insecure podman registry and then loading the ceph image to all hosts
+ {% if quick_install is defined %}
+ - echo -e "[registries.insecure]\nregistries = ['localhost:5000']" | sudo tee /etc/containers/registries.conf
+ - podman run -d -p 5000:5000 --name my-registry registry:2
+ - podman load -i /root/ceph_image.tar
+ - podman tag quay.ceph.io/ceph-ci/ceph:main localhost:5000/ceph
+ - podman push localhost:5000/ceph
+ - rm -f /root/ceph_image.tar
+ {% endif %}
```
Comment (epuertat) [inline src/pybind/mgr/dashboard/ci/cephadm/ceph_cluster.yml:50]: Same thing here.

File: src/pybind/mgr/dashboard/ci/cephadm/quick-bootstrap.sh [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    podman pull "${image_name}"
+fi
+
+rm -f ceph_image.tar
+
+printf "Saving the image: %s\n" "$image_name"
+podman save -o ceph_image.tar quay.ceph.io/ceph-ci/ceph:main
+
+printf "Creating the plan\n"
+kcli create plan -f ceph_cluster.yml -P ceph_dev_folder=/home/nia/projects/ceph -P quick_install=True -P expanded_cluster=True ceph
```
Comment (epuertat) [inline src/pybind/mgr/dashboard/ci/cephadm/quick-bootstrap.sh:52]: /home/nia/projects/ceph?

File: src/pybind/mgr/dashboard/ci/cephadm/start-cluster.sh [hunk 7]
```diff
@@ -59,8 +59,8 @@ fi
 npm run build ${FRONTEND_BUILD_OPTS} &
 
 cd ${CEPH_DEV_FOLDER}
-: ${VM_IMAGE:='fedora36'}
-: ${VM_IMAGE_URL:='https://archives.fedoraproject.org/pub/archive/fedora/linux/releases/36/Cloud/x86_64/images/Fedora-Cloud-Base-36-1.5.x86_64.qcow2'}
+: ${VM_IMAGE:='fedora40'}
+: ${VM_IMAGE_URL:='https://download.fedoraproject.org/pub/fedora/linux/releases/40/Cloud/x86_64/images/Fedora-Cloud-Base-Generic.x86_64-40-1.14.qcow2'}
```
Comment (afreen23) [inline src/pybind/mgr/dashboard/ci/cephadm/start-cluster.sh:63]: Is this required for the PR ?

File: src/pybind/mgr/dashboard/ci/cephadm/quick-bootstrap.sh (hunk 1 above)
Comment (nizamial09) [reply src/pybind/mgr/dashboard/ci/cephadm/quick-bootstrap.sh]: done, thanks

File: src/pybind/mgr/dashboard/ci/cephadm/quick-bootstrap.sh (hunk 2 above)
Comment (nizamial09) [reply src/pybind/mgr/dashboard/ci/cephadm/quick-bootstrap.sh:70]: i left it as it is since the kcli ip is currently static.

File: src/pybind/mgr/dashboard/ci/cephadm/ceph_cluster.yml (hunk 3 above)
Comment (nizamial09) [reply src/pybind/mgr/dashboard/ci/cephadm/ceph_cluster.yml]: thanks, i separated them out.

File: src/pybind/mgr/dashboard/ci/cephadm/ceph_cluster.yml (hunk 5 above)
Comment (nizamial09) [reply src/pybind/mgr/dashboard/ci/cephadm/ceph_cluster.yml]: done too

File: src/pybind/mgr/dashboard/ci/cephadm/bootstrap-cluster.sh (hunk 4 above)
Comment (nizamial09) [reply src/pybind/mgr/dashboard/ci/cephadm/bootstrap-cluster.sh]: done here

File: src/pybind/mgr/dashboard/ci/cephadm/quick-bootstrap.sh (hunk 6 above)
Comment (nizamial09) [reply src/pybind/mgr/dashboard/ci/cephadm/quick-bootstrap.sh]: removed now. getting the dir from a param called `-dir` or `--ceph-dir`

File: src/pybind/mgr/dashboard/ci/cephadm/start-cluster.sh (hunk 7 above)
Comment (nizamial09) [reply src/pybind/mgr/dashboard/ci/cephadm/start-cluster.sh:63]: it'd improve the speed so yeah, it is. and since I am upgrading the version in `ceph_cluster.yml` file, this partivular image should exist in the network pool we create in jenkins.

**Comment by nizamial09** [comment]
@Pegonzal @epuertat @afreen23 address your comments, please take a look again. thanks

**Review by nmunet — APPROVED** [review]
Tested locally, and it worked like a charm!

File: src/pybind/mgr/dashboard/ci/cephadm/quick-bootstrap.sh [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+for arg in "$@"; do
+  case "$arg" in
+    -u|--use-cached-image)
+      use_cached_image=true
+      ;;
+    -dir=*|--ceph-dir=*)
+      ceph_dev_folder=${arg#*=}
+      ;;
+    -e|--expanded-cluster)
+      expanded_cluster=true
```
[the remark this answers is not in the document]
Comment (nizamial09) [reply src/pybind/mgr/dashboard/ci/cephadm/quick-bootstrap.sh:32]: probably that involves changes in the [bootstrap-cluster.sh](https://github.com/ceph/ceph/blob/main/src/pybind/mgr/dashboard/ci/cephadm/bootstrap-cluster.sh#L39) file itself because as it is right now if you provide the `expanded_cluster` its going to deploy a fully functional expanded cluster with osds on it. 

If we want, later on we can introduce flags like `--deploy-all-hosts`, `--deploy-rgw` and so on. Or even look into something like the vstart does. `MGR=2 RGW=1 quick-bootstrap.sh`.

File: src/pybind/mgr/dashboard/ci/cephadm/quick-bootstrap.sh [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    fi
+done
+
+printf "Copying the image to the hosts\n"
+
+for node in $(seq 0 $((node_count - 1))); do
+    scp -o StrictHostKeyChecking=no ceph_image.tar root@192.168.100.10"${node}":/root/
+done
+
+rm -f ceph_image.tar
```
[the remark this answers is not in the document]
Comment (nizamial09) [reply src/pybind/mgr/dashboard/ci/cephadm/quick-bootstrap.sh:85]: Removing it from the host machine since its created by the script only to copy it inside the vms so that it can be loaded using podman. No longer needed for any other purposes.

File: src/pybind/mgr/dashboard/ci/cephadm/quick-bootstrap.sh [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+image_name=$(echo "$CEPHADM_IMAGE" | awk -F'/' '{print $NF}')
+ceph_cluster_yml='ceph_cluster.yml'
+node_count=$(awk '/nodes:/ {print $2}' "${ceph_cluster_yml}")
+
+if [[ ${use_cached_image} == false ]]; then
+    printf "Pulling the image: %s\n" "$image_name"
+    podman pull "${image_name}"
+fi
+
+rm -f ceph_image.tar
```
[the remark this answers is not in the document]
Comment (nizamial09) [reply src/pybind/mgr/dashboard/ci/cephadm/quick-bootstrap.sh:55]: Removing it from the host machine since its created by the script only to copy it inside the vms so that it can be loaded using podman. No longer needed for any other purposes.

here i am just making sure that it doesn't exist before.

File: src/pybind/mgr/dashboard/ci/cephadm/start-cluster.sh (hunk 7 above)
Comment (nizamial09) [reply src/pybind/mgr/dashboard/ci/cephadm/start-cluster.sh:63]: f36 is very old and we are pulling it from the archives which is even slower when downloading it. And refreshing the dependencies in f36 is significantly slower for me atleast and it was improved when I used f40. Which is why I went ahead with the change. If we want we can have it done separately but I didn't see the need for it as it is improving the speed for me locally. If its not the case in your testing, then I can remove this change and do it separately.

File: src/pybind/mgr/dashboard/ci/cephadm/quick-bootstrap.sh (hunk 8 above)
[the remark this answers is not in the document]
Comment (afreen23) [reply src/pybind/mgr/dashboard/ci/cephadm/quick-bootstrap.sh:32]: Okay even if I dont pass this flag I am getting all osds scheduled.
And after looking I see the bootsrap.sh just check whether this flag is defined, ehnce even if we pass `false` it goes on creating.

So we need to update that or not pass this option if falsy.

**Review by afreen23 — COMMENTED** [review]
Just one thingy, thanks for patience :)

**Review by afreen23 — APPROVED** [review]
